File: android/src/main/kotlin/com/loopit/minis/sys/Paths.kt
package com.loopit.minis.sys

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.UUID

class Paths(private val appContext: Context, messenger: BinaryMessenger) {
  private val channel = MethodChannel(messenger, CHANNEL)

  init {
    channel.setMethodCallHandler { call, result ->
      try {
        when (call.method) {
          "cacheDir" -> result.success(mapOf("path" to appContext.cacheDir.absolutePath))
          "appSupportDir" -> {
            val dir = File(appContext.filesDir, "support").apply { if (!exists()) mkdirs() }
            result.success(mapOf("path" to dir.absolutePath))
          }
          "documentsDir" -> result.success(mapOf("path" to appContext.filesDir.absolutePath))
          "externalDir" -> {
            val dir = appContext.getExternalFilesDir(null)
            result.success(mapOf("path" to (dir?.absolutePath)))
          }
          "tempFile" -> {
            val ext = call.argument<String>("ext")?.trimStart('.')?.lowercase() ?: "tmp"
            val name = "${UUID.randomUUID()}.$ext"
            val f = File(appContext.cacheDir, name)
            result.success(mapOf("path" to f.absolutePath))
          }
          "join" -> {
            val parts = call.argument<List<String>>("parts") ?: emptyList()
            val joined = parts.fold(File("")) { acc, p ->
              if (acc.path.isEmpty()) File(p) else File(acc, p)
            }.path
            result.success(mapOf("path" to joined))
          }
          "extension" -> {
            val p = call.argument<String>("path") ?: ""
            val idx = p.lastIndexOf('.')
            val sep = maxOf(p.lastIndexOf('/'), p.lastIndexOf('\\'))
            val ext = if (idx > sep && idx >= 0) p.substring(idx) else ""
            result.success(mapOf("ext" to ext))
          }
          "basename" -> {
            val p = call.argument<String>("path") ?: ""
            result.success(mapOf("name" to File(p).name))
          }
          else -> result.notImplemented()
        }
      } catch (t: Throwable) {
        result.error("paths_error", t.message, null)
      }
    }
  }

  fun dispose() { channel.setMethodCallHandler(null) }

  companion object { const val CHANNEL = "loopit/minis/paths" }
}

File: ios/Classes/Sys/Paths.swift
import Flutter
import Foundation

final class Paths {
  static let channelName = "loopit/minis/paths"
  private let channel: FlutterMethodChannel

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Paths.channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  func dispose() { channel.setMethodCallHandler(nil) }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let fm = FileManager.default
    switch call.method {
    case "cacheDir":
      let url = fm.urls(for: .cachesDirectory, in: .userDomainMask).first
      result(["path": url?.path as Any])
    case "appSupportDir":
      let url = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
      if let u = url, !fm.fileExists(atPath: u.path) {
        try? fm.createDirectory(at: u, withIntermediateDirectories: true)
      }
      result(["path": url?.path as Any])
    case "documentsDir":
      let url = fm.urls(for: .documentDirectory, in: .userDomainMask).first
      result(["path": url?.path as Any])
    case "externalDir":
      result(["path": NSNull()])
    case "tempFile":
      let args = call.arguments as? [String: Any]
      var ext = (args?["ext"] as? String ?? "tmp").lowercased()
      if ext.hasPrefix(".") { ext.removeFirst() }
      let dir = fm.urls(for: .cachesDirectory, in: .userDomainMask).first
      let url = dir?.appendingPathComponent("\(UUID().uuidString).\(ext)")
      result(["path": url?.path as Any])
    case "join":
      let args = call.arguments as? [String: Any]
      let parts = (args?["parts"] as? [String]) ?? []
      var joined = ""
      for p in parts {
        if joined.isEmpty { joined = p; continue }
        if joined.hasSuffix("/") || p.hasPrefix("/") { joined += p } else { joined += "/" + p }
      }
      result(["path": joined])
    case "extension":
      let args = call.arguments as? [String: Any]
      let p = (args?["path"] as? String) ?? ""
      let ns = (p as NSString)
      let ext = ns.pathExtension
      result(["ext": ext.isEmpty ? "" : ".\(ext)"])
    case "basename":
      let args = call.arguments as? [String: Any]
      let p = (args?["path"] as? String) ?? ""
      result(["name": (p as NSString).lastPathComponent])
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
